### PR TITLE
FIX: Post uploads setting access_control_post_id unnecessarily

### PR DESCRIPTION
**Pull request by martin-brennan**
This commit addresses an issue for sites where secure_uploads
is turned on after the site has been operating without it for
some time.

When uploads are linked when they are used inside a post,
we were setting the access_control_post_id unconditionally
if it was NULL to that post ID and secure_uploads was true.

However this causes issues if an upload has been used in a
few different places, especially if a post was previously
used in a PM and marked secure, so we end up with a case of
the upload using a public post for its access control, which
causes URLs to not use the /secure-uploads/ path in the post,
breaking things like image uploads.

We should only set the access_control_post_id if the post is the first time the
upload is referenced so it cannot hijack uploads from other places.

### CODE REVIEW

File: app/models/post.rb [hunk 1]
```diff
@@ -1093,11 +1093,30 @@ def link_post_uploads(fragments: nil)
...
+            .where(id: upload_ids, access_control_post_id: nil)
+            .where("uploads.id NOT IN (SELECT upload_id FROM custom_emojis)")
+
+        access_control_will_change_upload_ids =
+          uploads_in_post
+            .map do |upl|
+              first_ref = upl.upload_references.sort_by { |ur| [ur.created_at, ur.id] }.first
+              upl.id if first_ref.blank? || first_ref.targets?(self)
+            end
+            .compact
```
Comment (Drenmi) [inline app/models/post.rb:1115]: We can use `#filter_map` and `#min_by` to make it slightly more succinct. 🙂 

```suggestion
          uploads_in_post
            .filter_map do |upl|
              first_ref = upl.upload_references.min_by { |ur| [ur.created_at, ur.id] }
              upl.id if first_ref.blank? || first_ref.targets?(self)
            end
```

File: app/models/upload_reference.rb [hunk 2]
```diff
@@ -6,6 +6,10 @@ class UploadReference < ActiveRecord::Base
 
   delegate :to_markdown, to: :upload
 
+  def targets?(target_to_check)
```
Comment (Drenmi) [inline app/models/upload_reference.rb:9]: Wondering if we should name this `#target?` to follow Ruby conventions, e.g. `Enumberable#include?`. Avoids ambiguity of whether "targets" is plural of "target".

**Review by Drenmi — APPROVED** [review]
I am happy to report that I understood the PR description. I think I'm starting to get buried deeper into the secure uploads quicksand. 🙂

File: app/models/upload_reference.rb (hunk 2 above)
Comment (martin-brennan) [reply app/models/upload_reference.rb:9]: Yes okay I will change :+1:

File: app/models/post.rb (hunk 1 above)
Comment (martin-brennan) [reply app/models/post.rb:1115]: Ah yes `filter_map`, the one I always recommend to others but forget myself :laughing: And thanks for `min_by` I didn't know that!

**Comment by martin-brennan** [comment]
@Drenmi good to hear your brain is adapting to the insanity of secure uploads nicely, you'll be in the asylum soon enough

**Pull request merged**